### ISSUE TITLE
tf.reduce_sum is hard to use in a numerical stable way

### ISSUE THREAD

**Issue by nfelt** [tensorflow/tensorflow#51419]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

### Edited to add:

As 1st comment clarifies, the described behavior is probably "working as intended" due to accumulation of floating point errors, which is mitigated in many common cases but not all cases, leading to surprising inconsistency.

However, I still think at least 2 things could be improved:

- `tf.reduce_sum` could include a note in its documentation explaining this behavior, like [`np.sum`](https://numpy.org/doc/stable/reference/generated/numpy.sum.html)
- `tf.reduce_sum` could take a `dtype` parameter to improve precision of the accumulator and output without needing to increase precision of the entire input tensor, also like [`np.sum`](https://numpy.org/doc/stable/reference/generated/numpy.sum.html)

Original report follows.

---

**System information**
This bug is reproducible on standard Colab, so I've only included the following as system information:
- TensorFlow version: 2.5.0
- Python version: 3.7.11

**Code to reproduce**

```python
with tf.device("cpu:0"):
  print(tf.reduce_sum(tf.ones((20_000_000, 2)), axis=0))
```

Full reproduction in colab on a GPU-enabled VM: https://gist.github.com/nfelt/e6edc45736740ed899eea95a36bea359

**Describe the current behavior**

On CPU, `tf.reduce_sum` on a rank-2 float32 tensor, along an axis containing more than 2^24 elements of value 1.0, emits an incorrect output that is truncated to 2^24 (the maximum integer precisely representable in float32).  I.e. the above code emits `tf.Tensor([16777216. 16777216.], shape=(2,), dtype=float32)`.

**Describe the expected behavior**

It should emit the same answer that it does in pretty much every other context, which is the numerically correct answer up to floating point representation error, in this case `tf.Tensor([20000000. 20000000.], shape=(2,), dtype=float32)`.

In particular, we get the expected answer in all of the following situations:
- Running `reduce_sum` on GPU instead of CPU
- Using `float64` instead of `float32`
- Summing a rank-1 tensor of more than 2^24 elements of value 1.0
- Summing a rank-2 tensor along an axis with fewer than 2^24, but individually larger, elements, e.g. `tf.reduce_sum(tf.ones((10_000_000, 2)) * 2.0, axis=0)`
  
**Notes**

It seems like the problem of stopping at 2^24 is affecting the count of how many elements get summed, rather than the actual sum itself, based on these cases:

```python
with tf.device("cpu:0"):
  print(tf.reduce_sum(tf.ones((20_000_000, 2)) * 0.5, axis=0))
>>> tf.Tensor([8388608. 8388608.], shape=(2,), dtype=float32)
```

In particular, it seems to only include the trailing 2^24 elements, based on this:

```python
tensor_with_one_big_value = tf.concat([[[10_000_000, 0]], tf.ones((20_000_000, 2))], axis=0)
with tf.device("cpu:0"):
  print(tf.reduce_sum(tensor_with_one_big_value, axis=0))
with tf.device("cpu:0"):
  print(tf.reduce_sum(tf.reverse(tensor_with_one_big_value, axis=[0]), axis=0))
>>> tf.Tensor([16777216. 16777216.], shape=(2,), dtype=float32)
>>> tf.Tensor([26777216. 16777216.], shape=(2,), dtype=float32)
```

**Comment by HVoltBb** [comment]
that's floating point arithmetic. adding 1 to 16777216 in float32 doesn't do anything, since 16777217 is not in float32. See [here](https://stackoverflow.com/questions/12596695/why-does-a-float-variable-stop-incrementing-at-16777216-in-c)

**Comment by nfelt** [comment]
@HVoltBb thanks for pointing that out, I guess you're right that this explains the behavior.  I still think it's quite surprising that it's so inconsistent - between CPU and GPU, and between rank-1 and rank-2 `tf.reduce_sum()` invocations - but looking at the [numpy docs](https://numpy.org/doc/stable/reference/generated/numpy.sum.html) I see it has a similar caveat that the choice of axis for the summation affects whether it uses a partial sum optimization to preserve better numerical accuracy in the overall results.  It would sure be nice if `tf.reduce_sum` documentation had an equivalent note, but fair enough that it's more a doc fix request than a bug fix then.

**Comment by HVoltBb** [comment]
@nfelt I assume when the job is offloaded to GPU, the underlying algorithm is breaking up the sum into multiple chunks, implicitly achieving a higher precision for this operation, and thus avoiding this issue. I haven't looked at rank-1 vs rank-2 difference, but will do it later.

**Comment by nfelt** [comment]
FWIW, one other thing that occurred to me is that `np.sum` supports a `dtype` parameter for the accumulator and result, while `tf.reduce_sum` does not, forcing a cast of the full input tensor to float64 to avoid this issue.  Unless there's some fancy lazy behavior happening under the hood or something to defer the cast, that would seem to require incurring a lot more memory overhead compared to numpy's approach.  So maybe this could also be reframed as an FR for adding that parameter.

> I haven't looked at rank-1 vs rank-2 difference, but will do it later.

Based on the note for `np.sum` about the partial sum optimization being axis-specific, I'm assuming the rank-1 vs rank-2 difference is actually a memory-aligned-axis difference, since from my testing, in the rank-2 case the effect only happens when reducing along axis 0, which I believe would be the non-memory-aligned direction.

**Comment by HVoltBb** [comment]
@nfelt  I think your explanation for the rank-1 vs rank-2 could be the case here.

**Comment by HVoltBb** [comment]
I came across this solution by accident today. The solution is the so called compensated sum, and it has been already implemented in tensorflow_probability. You can use the following code snippet to get the correct answer.
```
import tensorflow_probability as tfp

with tf.device("cpu:0"):
  print(tfp.math.reduce_kahan_sum(tf.ones((20_000_000, 2)), axis=0))
```

**Comment by epbsb** [comment]
I was using tf.reduce_sum with complex64 dtype to sum a lot of numbers (1400x1400 matrix), and I was amazed to see how different the results are from cpu and from executing the numpy sum, the cpu results are totally wrong. In the GPU the bigger diference is on the complex result, while the real part is off by 30%. Honestly, I don't know how it can be so wrong. I'm assuming the numpy answer is the closest to the right one by the way, since it has been more tested and implemented.
